libraries/AP_UAVCAN/UAVCAN_UARTDriver.cpp

```diff
@@ -0,0 +1,168 @@
+#include "UAVCAN_UARTDriver.h"
+
+
+#define AP_UAVCAN_UARTDRIVER_BUF_SIZE_MIN_TX            (512*4)
+#define AP_UAVCAN_UARTDRIVER_BUF_SIZE_MIN_RX            (512*4)
+
+extern const AP_HAL::HAL& hal;
+
+UAVCAN_UARTDriver::UAVCAN_UARTDriver()
+{
+    _write_mutex = hal.util->new_semaphore();
+    _read_mutex = hal.util->new_semaphore();
+}
+
+/*
+  open virtual port
+ */
+void UAVCAN_UARTDriver::begin(uint32_t baud)
+{
+    begin(baud, 0, 0);
+}
+
+void UAVCAN_UARTDriver::begin(uint32_t baud, uint16_t rxS, uint16_t txS)
+{
+    (void)baud; // unused in virtual ports
+
+    _initialised = false;
+
+    // enforce minimum sizes
+    if (rxS < AP_UAVCAN_UARTDRIVER_BUF_SIZE_MIN_RX) {
+        rxS = AP_UAVCAN_UARTDRIVER_BUF_SIZE_MIN_RX;
+    }
+    if (txS < AP_UAVCAN_UARTDRIVER_BUF_SIZE_MIN_TX) {
+        txS = AP_UAVCAN_UARTDRIVER_BUF_SIZE_MIN_TX;
+    }
+
+    if (_writebuf.set_size(txS) && _readbuf.set_size(rxS)) {
+        _initialised = true;
+    }
+}
+
+
+/*
+  shutdown a Virtual UART
+ */
+void UAVCAN_UARTDriver::end()
+{
+    _initialised = false;
+    _readbuf.set_size(0);
+    _writebuf.set_size(0);
+}
+
+
+/*
+  do we have any bytes pending transmission?
+ */
+bool UAVCAN_UARTDriver::tx_pending()
+{
+    // any data in SerialManager -> UAVCAN
+    return (_writebuf.available() > 0);
+}
+
+/*
+  return the number of bytes available to be read
+ */
+uint32_t UAVCAN_UARTDriver::available()
+{
+    if (!_initialised) {
+        return 0;
+    }
+    // any data in UAVCAN -> SerialManager
+    return _readbuf.available();
+}
+
+/*
+  how many bytes are available in the output buffer?
+ */
+uint32_t UAVCAN_UARTDriver::txspace()
+{
+    if (!_initialised) {
+        return 0;
+    }
+    // how much free space available in SerialManager -> UAVCAN
+    return _writebuf.space();
+}
+
+uint32_t UAVCAN_UARTDriver::handle_inbound(const uint8_t *buffer, uint32_t size)
+{
+    if (!_read_mutex->take(2)) {
+        return 0;
+    }
+
+    // load data into UAVCAN -> SerialManager
+    uint32_t len = _readbuf.write(buffer, size);
+    _read_mutex->give();
+    return len;
+}
+
+int16_t UAVCAN_UARTDriver::fetch_for_outbound(void)
+{
+    if (!_initialised) {
+        return -1;
+    }
+    if (!_write_mutex->take(2)) {
```

Comment:
we don't need mutexes on the readbuf and writebuf (they are lockless data structures, as long as there is only one reader and one writer)

Reply:
are you sure? I think I was having a higher byte drop rate before I added these.. which is why I added them. I didn't document the change so I can't confirm that though. Seemed like a reasonable thing to do though.